csharp/src/Google.Protobuf/CodedOutputWriter.cs

```diff
@@ -0,0 +1,589 @@
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#endregion
+
+#if NETSTANDARD2_0
+using System;
+using System.Buffers;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Google.Protobuf
+{
+    /// <summary>
+    /// Encodes and writes protocol message fields.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// This class is generally used by generated code to write appropriate
+    /// primitives to the stream. It effectively encapsulates the lowest
+    /// levels of protocol buffer format. Unlike some other implementations,
+    /// this does not include combined "write tag and value" methods. Generated
+    /// code knows the exact byte representations of the tags they're going to write,
+    /// so there's no need to re-encode them each time. Manually-written code calling
+    /// this class should just call one of the <c>WriteTag</c> overloads before each value.
+    /// </para>
+    /// <para>
+    /// Repeated fields and map fields are not handled by this class; use <c>RepeatedField&lt;T&gt;</c>
+    /// and <c>MapField&lt;TKey, TValue&gt;</c> to serialize such fields.
+    /// </para>
+    /// </remarks>
+    public ref struct CodedOutputWriter
+    {
+        // "Local" copy of Encoding.UTF8, for efficiency. (Yes, it makes a difference.)
+        internal static readonly Encoding Utf8Encoding = Encoding.UTF8;
+
+        private BufferWriter writer;
+        private Encoder encoder;
+
+        public CodedOutputWriter(IBufferWriter<byte> writer)
+        {
+            this.writer = new BufferWriter(writer);
+            this.encoder = null;
+        }
+
+        #region Writing of values (not including tags)
+
+        /// <summary>
+        /// Writes a double field value, without a tag, to the stream.
+        /// </summary>
+        /// <param name="value">The value to write</param>
+        public void WriteDouble(double value)
+        {
+            WriteRawLittleEndian64((ulong)BitConverter.DoubleToInt64Bits(value));
+        }
+
+        /// <summary>
+        /// Writes a float field value, without a tag, to the stream.
+        /// </summary>
+        /// <param name="value">The value to write</param>
+        public void WriteFloat(float value)
+        {
+            // netstandard2.0 target does not provide a non-allocating API for converting float values
+            byte[] rawBytes = BitConverter.GetBytes(value);
+            if (!BitConverter.IsLittleEndian)
+            {
+                ByteArray.Reverse(rawBytes);
+            }
+
+            writer.Write(rawBytes);
```

Comment:
You could try to stackalloc a span and write to that, OR, create a `Write(float)` method on the `BufferWriter` itself (much faster, no copying required):
```C#
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void Write(float value)
        {
            Span<byte> destination = GetSpan(sizeof(float));

            Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(destination), value);

            if (!BitConverter.IsLittleEndian)
            {
                destination.Slice(0, sizeof(float)).Reverse();
            }

            Advance(4);
        }
```

```C#
// Might be slower than what you have currently (due to stackalloc-zero init/etc.), but at least it doesn't allocate.
public void WriteFloat(float value)
{
    Span<byte> floatSpan;
    // Cannot create a span directly since it gets passed to instance methods on a ref struct.
    unsafe
    {
        byte* ptr = stackalloc byte[sizeof(float)];
        floatSpan = new Span<byte>(ptr, sizeof(float));
    }

    Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(floatSpan), value);

    if (!BitConverter.IsLittleEndian)
    {
        floatSpan.Reverse();
    }

    writer.Write(floatSpan);
}
```